cni/network/multitenancy.go

```diff
@@ -202,31 +203,39 @@ func (m *Multitenancy) GetAllNetworkContainers(
 
 	ncResponses, hostSubnetPrefixes, err := m.getNetworkContainersInternal(ctx, podNamespace, podNameWithoutSuffix)
 	if err != nil {
-		return []IPAMAddResult{}, fmt.Errorf("%w", err)
+		return IPAMAddResult{}, fmt.Errorf("%w", err)
 	}
 
 	for i := 0; i < len(ncResponses); i++ {
 		if nwCfg.EnableSnatOnHost {
 			if ncResponses[i].LocalIPConfiguration.IPSubnet.IPAddress == "" {
 				logger.Info("Snat IP is not populated for ncs. Got empty string",
 					zap.Any("response", ncResponses))
-				return []IPAMAddResult{}, errSnatIP
+				return IPAMAddResult{}, errSnatIP
 			}
 		}
 	}
 
-	ipamResults := make([]IPAMAddResult, len(ncResponses))
+	ipamResult := IPAMAddResult{}
+	ipamResult.interfaceInfo = make(map[string]network.InterfaceInfo)
 
 	for i := 0; i < len(ncResponses); i++ {
-		ipamResults[i].ncResponse = &ncResponses[i]
-		ipamResults[i].hostSubnetPrefix = hostSubnetPrefixes[i]
-		ipconfig, routes := convertToIPConfigAndRouteInfo(ipamResults[i].ncResponse)
-		ipamResults[i].defaultInterfaceInfo.IPConfigs = []*network.IPConfig{ipconfig}
-		ipamResults[i].defaultInterfaceInfo.Routes = routes
-		ipamResults[i].defaultInterfaceInfo.NICType = cns.InfraNIC
+		// one ncResponse gets you one interface info in the returned IPAMAddResult
+		ifInfo := network.InterfaceInfo{
+			NCResponse:       &ncResponses[i],
+			HostSubnetPrefix: hostSubnetPrefixes[i],
+		}
+
+		ipconfig, routes := convertToIPConfigAndRouteInfo(ifInfo.NCResponse)
+		ifInfo.IPConfigs = []*network.IPConfig{ipconfig}
+		ifInfo.Routes = routes
+		ifInfo.NICType = cns.InfraNIC
+
+		// assuming we only assign infra nics in this function
+		ipamResult.interfaceInfo[string(ifInfo.NICType)+strconv.Itoa(i)] = ifInfo
```

Comment:
I think this might be the reason for having it `string` and not `NICType`. 
Might be useful to have a function that returns this `key` if we need to access this in multiple places. Otherwise, we will need to change the key everywhere if this changes in the future.

Reply:
added a getInterfaceInfoKey method to each of the 3 possible ways we populate ipam addresults/interface info:
- multitenancy
- cns invoker
- azure invoker
Having one unified function for all the different methods would be strange because, for example, in multitenancy, we input an index local to a loop. Once the map is populated in each of these 3 cases, I assume that we never access the map by key again (we only loop over its contents).

Reply:
The three key formats are:
- Multitenancy: nic type + index in for loop
- CNS Invoker: Either mac address (if delegated nic) or infra nic (if infra nic)
- Azure Invoker: Always infra nic